File: algorithm/188. Best Time to Buy and Sell Stock IV/cpp/BestTimetoBuyandSellStockIV.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        int n = prices.size();
        if(n < 2) return 0;
        vector<vector<vector<int>>> dp(2,vector<vector<int>>(k+1, vector<int>(n)));
        for(int i = 0; i < k; i++){
            dp[1][i][0] = -prices[0];
        }
        for(int i = 1; i < n; i++){
            for(int j = 0; j < k+1; j++){
                if(j!=0) dp[0][j][i] = max(dp[0][j][i-1], dp[1][j-1][i-1]+prices[i]);
                dp[1][j][i] = max(dp[1][j][i-1], dp[0][j][i-1]-prices[i]);
            }
        }
        return dp[0][k][n-1];
    }
};
/*
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        int n = prices.size();
        if(n < 1) return 0;
        vector<vector<vector<int>>> dp(2,vector<vector<int>>(k+1, vector<int>(n)));
        for(int i = 0; i < k; i++){
            dp[1][i][0] = -prices[0];
        }
        for(int i = 1; i < n; i++){
            for(int j = 0; j < k+1; j++){
                if(j!=0) dp[0][j][i] = max(dp[0][j][i-1], dp[1][j-1][i-1]+prices[i]);
                dp[1][j][i] = max(dp[1][j][i-1], dp[0][j][i-1]-prices[i]);
            }
        }
        return dp[0][k][n-1];
    }
};
*/

File: algorithm/188. Best Time to Buy and Sell Stock IV/java/BestTimetoBuyandSellStockIV.java
// DP
// using idea of state machine
class Solution {
    public int maxProfit(int k, int[] prices) {
        int n = prices.length;
        if(n < 2) return 0;
        int[][][] dp = new int[2][k+1][n];
        for(int i = 0;  i < k+1; i++){
            dp[1][i][0] = -prices[0];
        }
        for(int i = 1; i < n; i++){
            dp[1][0][i] = Math.max(dp[1][0][i-1], dp[0][0][i-1]-prices[i]);
            for(int j = 1; j < k+1; j++){
                dp[0][j][i] = Math.max(dp[0][j][i-1], dp[1][j-1][i-1]+prices[i]);
                dp[1][j][i] = Math.max(dp[1][j][i-1], dp[0][j][i-1]-prices[i]);
            }
        }
        return dp[0][k][n-1];
    }
}
